### ISSUE TITLE
Test Runner

### ISSUE THREAD

**Issue by lukeapage** [microsoft/playwright#654]
Hi,

I have been working on a POC for a test runner based on top of mocha that uses playwright, with the following features:

 - retryability e.g. if you expect an element to have text 'x' it will retry the get elements and getting text until 'x' is there
 - jest style matchInlineSnapshot
 - a UI to allow selecting test files and running only a specific test
 - Synchronous execution (no async/await - allows for better UX in a UI, allows retryability and simpler looking tests, with the downside of limited value for the end user in node debugging, made up for by a step-through UI) - similar to cypress
 - working with playwright in a parallel manner

I have more info but I don't want to publish my private repo until its a bit more complete.

and I wanted to know if any of this overlaps with direct plans for playwright or perhaps overlaps with another tool your aware of - I have no desire to try and create a competitor its just it seems a test runner that brings together the best of playwright, jest and mocha is somewhat lacking.

Specifically, given that I guess you could use jest and playwright together today whats missing is retryability, mixed with assertion and snapshots.

**Comment by pavelfeldman** [comment]
Playwright is an automation driver and we hope it fits the existing ecosystem of the test runners, including mocha, jest, etc. It aims at being a driver specifically for testing, so we are happy to include capabilities that make test runners integrations' lives easier.

> retryability e.g. if you expect an element to have text 'x' it will retry the get elements and getting text until 'x' is there

Since we already have the retry logic for waiting for element to be in DOM, be visible, clickable, etc, we could extend `page.waitForSelector` with the `predicate` function option for this kind of retrying expect. So feel free to file an issue or send a PR for it in Playwright.

> jest style matchInlineSnapshot

Here as well, if you think there is something we could do in Playwright that would be universally helpful for all the test runners, file an issue!

> Synchronous execution (no async/await)

Thinking out loud on this one. That's totally up to you. We ourselves chose async/await as a compromise between capabilities and readability. We believe that's the most ergonomically sound way of expressing driver capabilities going forward. And we hope test runners integrate Playwright along with its API spirit - that makes it easier for the users to learn scripting patterns only once.

Also, as a heads up, since we endorse no-setTimeout testing, we are going to provide more and more `waitFor` signals for test script to know when the action is complete. The current pattern for it is something like this:

```js
await Promise.all([
  page.waitForFoo(),
  page.waitForBar(),
  page.click('selector');
]);
```

It is somewhat hairy, but it allows for great flexibility. I'm not sure it will be easy to express extensible structures like this in a sync manner.

> and I wanted to know if any of this overlaps with direct plans for playwright or perhaps overlaps with another tool your aware of

This is not something we are currently looking at, our primary focus is the driver. Let's leave this Issue open and see if others had similar ideas / plans.

> Specifically, given that I guess you could use jest and playwright together today whats missing is retryability, mixed with assertion and snapshots.

We welcome all the test runners and will accept requests for better integration - retryability, snapshotting, assertions all sound like good candidates - whatever makes sense as a part of the driver itself for efficiency. With that, it'll eventually make jest+playwright story very close to what you are describing if I understand it correctly.

**Comment by lukeapage** [comment]
Thanks for the detailed response, thats what I was looking for. Sorry for my long reply.

To add where I am coming from - I've led a team of 70 developers that wrote 2000 integration tests for a large and complex web app - I've personally tracked down problems causing inconsistent tests until we got to the point where we have a 99.99% pass rate. We did this using cypress.

> Since we already have the retry logic for waiting for element to be in DOM, be visible, clickable, etc, we could extend page.waitForSelector with the predicate function option for this kind of retrying expect. So feel free to file an issue or send a PR for it in Playwright.

Thats great, but I am a little hesitant as to whether these kind of problems can and should be solved by something like playwright. The reason is that it solves some of the problems, but not all the problems.

For instance..

```
const component  = await page.$('.component');
const subComponent = await component.$('.sub');
expect(await subComponent.evaluate((el) => el.innerText)).toMatchInlineSnapshot();
```

In a simple application, waiting for the elements to exist may be enough e.g.

```
const component  = await page.waitForSelector('.component');
const subComponent = await component.waitForSelector('.sub');
expect(await subComponent.evaluate((el) => el.innerText)).toMatchInlineSnapshot();
```

but, what if the innerText starts with 'Loading...'. It seems you have two choices - adding application logic to append a 'test-loaded' class which you can wait for, or putting the expected text value into the assertion so that you end up with..

```
const component  = await page.waitForSelector('.component');
await component.waitForSelector('.sub', (el) => el.innerText === 'expected value');
```

and then relying on playwright to throw if the assertion fails. essentially the more you go down this road, the more playwright becomes its own assertion framework.

and.. you get issues if for instance the parent component might for whatever reason get removed from the dom during the re-render.. it pushes you to try and put the whole get and assert all into a single statement and imo it starts to get messy.

Don't get me wrong - I think having waitFor type functions is very useful, I just think that in an integration test it doesn't solve all the problems. As an end user I would like my testers and developers to be able to write the simplest integration tests, have them be reliable without having to debug why they sometimes fail and have the api look as simple as possible.

>> jest style matchInlineSnapshot
> Here as well, if you think there is something we could do in Playwright that would be universally helpful for all the test runners, file an issue!

The reason why jest wouldn't work out of the box (as above) is because of waiting - to use jest as-is you would have to always wait for some other condition before going forward with the assertion.

I don't think it belongs inside playwright - I think it belongs on the test runner side, because its to do with the test file itself and managing the test run (for instance you wouldn't want to write the snapshots until all tests were complete).

> Thinking out loud on this one. That's totally up to you. We ourselves chose async/await as a compromise between capabilities and readability. We believe that's the most ergonomically sound way of expressing driver capabilities going forward. And we hope test runners integrate Playwright along with its API spirit - that makes it easier for the users to learn scripting patterns only once.

This is what I have the biggest doubts about - I am not a fan of magic. It comes down to retryability and bridging the gap between test automation that you've done and test-runner/assertion.

What happens if I do this?
```
const button  = await page.$(`.button`);
await button.click();
```
and the button element is removed from the dom and re-added with one with the same selector? This comes up surprisingly often with frameworks like react.

> Also, as a heads up, since we endorse no-setTimeout testing, we are going to provide more and more `waitFor` signals for test script to know when the action is complete.

I am completely for that - setTimeout testing is the death and our example 2000 tests above don't have any waits in at all. What I did for our test application was to proxy `fetch`, `mutationObserver`, `setTimeout`, `setInterval` with a class that recorded if a timeout was waiting to occur. Then before getting every element I made sure we waited for all activity to stop. This worked well - the only thing I couldn't get to work was MutationObserver because I couldn't wrap every dom operation that might trigger a MutationObserver. The only caveat was that I only wait for timeouts under a certain number - because a timeout for 100ms might be worth waiting for as its a debounce, but 2000ms might be a user timeout that we shouldn't wait for.

Do you plan on adding `waitFor` that allows covering timers like the above? That would be amazing, though essentially I'd want to wait for all of that after any action to get maximum reliability for free.

on that subject, we also use lolex to mock time in the application so that 1. if for instance an operation times out after 7.5 seconds and we are running it on a slow cloud machine, we dont get occasional problems 2. so that if we want to test a timeout we can forward time rather than waiting. Are there plans around this for Playwright?

**Comment by pavelfeldman** [comment]
Thanks for the thorough analysis!

>  the more playwright becomes its own assertion framework.

There is no intention to become our own assertion framework. We'd like to include features that are more efficient if they are implemented inside the driver, that's it. Scenarios of swapping nodes that match the same selector, react rendering that changes node identity, etc are very generic these days, so addressing them seems logical. If we are already polling rafs, we can include a user predicate into the mix and let them wait for 'Loading...' text to change. The rationale is that the higher level of abstraction we use to address those, the less efficient solution would be in terms of CPU cycles.

> I don't think it belongs inside playwright - I think it belongs on the test runner side, because its to do with the test file itself and managing the test run

I did not mean that Playwright would want to do any of test runner business. What I meant was that we might be more efficient at collecting the snapshot or dynamically matching it, so we are open to requests in that area.

> What happens if I do this ... and the button element is removed from the dom and re-added with one with the same selector?

I don't think disappearing button has anything to do with the sync / async call conventions. We don't recommend using element handles unless necessary, we would always recommend doing something like this:

```js
await page.click(`.button`);
```

We've also been thinking about introducing a `SelectorHandle` that would be dynamically resolved from a selector at the time of use. That way even your code snippet would work. More importantly, we'll probably never see awaiting for the click, as you say, we would want to see a signal beside it that would unlock assertion.

> What I did for our test application was to proxy...

I know what you are talking about, achieving determinism is handy when fighting flakiness. Chromium even has a concept of virtual time that you can use at own risk. We gave up on determinism though - in the modern web, too many decisions are made at runtime, based on real-time latencies and flips of the coin. So we can't IO to line up the same way on the timeline.

> Then before getting every element I made sure we waited for all activity to stop.

Similarly to the above, on the real world sites activity never stops. There are always pending analytics requests, ever-rescheduling timers, etc. In reality, we have very little idea about layout stability, idleness detection, etc even inside the browser. So we would try as much as we can to not fall back to the idleness heuristics and give users enough signals to reliably know that the test step is complete and is ready for the assertion. We might give up eventually, not for the lack of trying!

> Do you plan on adding waitFor that allows covering timers like the above? That would be amazing, though essentially I'd want to wait for all of that after any action to get maximum reliability for free.

It is hard to distinguish noise from signal in the post-action activity. We've seen how click produced a stray analytics long poll that never ended, until the next action. With that, we'd much rather invest into firing enough signals for you to consider that the next action can be performed. Triggers that would fire upon either IO or DOM changes.

> On a slow cloud machine, we dont get occasional problems & Fast forward time rather than waiting. Are there plans around this for Playwright?

Not at the moment. The reason is the same - we can't virtualize time w/o aligning it with the raf sequence and animation timeline. Only mocking time breaks combined css+js animations, etc. We have a powerful `virtual time` concept in Chromium, even exposed over the protocol with the manual time budgeting and frame control. It aligns animations, io, time, etc. But even within the browser the implementation was not bullet-proof. It fails on random renderer->browser calls and auto-advances the time eagerly. So we don't have too many good ideas on how this could be achieved.

**Comment by lukeapage** [comment]
Thanks, thats been helpful.

> Similarly to the above, on the real world sites activity never stops

We did manage to solve this, but we made a special build that 1) in a few small places acts differently (so if window.AUTOMATED_TEST is defined for instance, it has longer timeouts and no animations) 2) we mocked out the websocket and all requests so that we wouldn't get data updates when we didn't want them.

I'll post a link to what I'm working on when its nearer an early version.
